consensus/propagation/have_wants.go

```diff
@@ -63,58 +66,143 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
+	}
+
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
+}
+
+// perPeerPerBlockConcurrentRequestLimit returns the maximum number of requests for a given block by peer.
+// it helps avoid downloading the whole block from a single peer.
+func perPeerPerBlockConcurrentRequestLimit() int64 {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var requestsBA *bits.BitArray
+	height := int64(0)
+	round := int32(0)
+	periodicRequestsDelay := 10 * time.Millisecond
+	periodicRequestTicker := time.NewTicker(periodicRequestsDelay)
+	for {
+		select {
+		case <-blockProp.ctx.Done():
+			return
+		case <-periodicRequestTicker.C:
+			// periodically send the existing requests
+			if requestsBA != nil && !requestsBA.IsEmpty() {
+				blockProp.sendWant(ps, height, round, requestsBA)
+				requestsBA = nil
+			}
+		case req, ok := <-ps.requestChan:
+			if !ok {
+				return
+			}
+
+			// concurrent per-peer per-block request limit reached
+			if ps.requestCount.Load() >= perPeerPerBlockConcurrentRequestLimit() {
+				// we reached the limit of requests, we request what we can
+				if requestsBA != nil && !requestsBA.IsEmpty() {
+					blockProp.sendWant(ps, height, round, requestsBA)
+					requestsBA = nil
+				}
+				// wait for a part to be received before continuing
+				select {
+				case <-blockProp.ctx.Done():
+					return
+				case _, ok = <-ps.receivedPart:
+					if !ok {
+						return
+					}
+				}
+			}
+
+			// if we're at a new height, we can drop the previous requests
+			if height != req.height || (height == req.height && round != req.round) {
+				requestsBA = nil
+				height = req.height
+				round = req.round
+			}
+
+			partIndex := req.index
+			_, parts, fullReqs, has := blockProp.getAllState(height, round, false)
+			if !has {
+				blockProp.Logger.Error("couldn't find proposal when filtering requests", "height", height, "round", round)
+				continue
+			}
+			if parts.BitArray().GetIndex(int(partIndex)) {
+				continue
+			}
+
+			reqLimit := 1
+			reqs := blockProp.countRequests(height, round, int(partIndex))
+			if len(reqs) >= reqLimit {
+				fullReqs.SetIndex(int(partIndex), true)
+				continue
+			}
+
+			// don't request the part from this peer if we've already requested it
+			// from them.
+			for _, p := range reqs {
+				// p == peer means we have already requested the part from this peer.
+				if p == ps.peer.ID() {
+					continue
+				}
+			}
+
+			if requestsBA == nil {
+				requestsBA = bits.NewBitArray(int(parts.Total()))
+			}
+			requestsBA.SetIndex(int(partIndex), true)
+
+			if len(requestsBA.GetTrueIndices()) >= wantBatchSize(int(parts.Total())) {
+				// no need to keep holding to a large number of requests
+				blockProp.sendWant(ps, height, round, requestsBA)
+				requestsBA = nil
```

Comment:
are we not sending a want until we have batch size? 

if so, I think a superior way to batch is to constantly clear the haves from this peer until either the channel is empty or our request limit is reached. That way we're constantly sending wants, yet batching when we can. This what I had in mind with batching before at least.

Reply:
> yet batching when we can

can you expand on this? how to know that we can batch? when receiving a have, there is no way to know that we will receive another have unless we add some other notification mechanism, no?

Reply:
> there is no way to know that we will receive another have unless we add some other notification mechanism,

we can check channel size after reading from the channel

Reply:
so we reap as much as we can from the channel until we hit the request limit. if we can do that, we're batching as much as we can while also not relying on a hot loop

Reply:
8c875a002421a50e8fe30a6ad203755fb0389324